README.md
= paperclip-azure

home  :: https://github.com/supportify/paperclip-azure
code  :: https://github.com/supportify/paperclip-azure
rdoc  :: https://www.rubydoc.info/docs/paperclip-azure
bugs  :: https://github.com/supportify/paperclip-azure/issues

== DESCRIPTION:

Paperclip-Azure is a [Paperclip](https://github.com/thoughtbot/paperclip) storage driver for storing files in a Microsoft Azure Blob.

== FEATURES/PROBLEMS:

* FIX (list of features or problems)

== SYNOPSIS:

The Azure storage engine has been developed to work as similarly to S3 storage configuration as is possible.  This gem can be configured in a Paperclip initializer or environment file as follows:

    Paperclip::Attachment.default_options[:storage] = :azure
    Paperclip::Attachment.default_options[:azure_credentials] = {
        storage_account_name: ENV['AZURE_STORAGE_ACCOUNT'],
        storage_access_key:   ENV['AZURE_STORAGE_ACCESS_KEY'],
        container:            ENV['AZURE_CONTAINER_NAME']
    }

Or, at the level of the model such as in the following example:

    has_attached_file :download,
                      storage: :azure,
                      azure_credentials: {
                        storage_account_name: ENV['AZURE_STORAGE_ACCOUNT'],
                        storage_access_key:   ENV['AZURE_STORAGE_ACCESS_KEY'],
                        container:            ENV['AZURE_CONTAINER_NAME']
                      }

Additionally, you can also supply credentials using a path or a File that contains the +storage_access_key+ and +storage_account_name+ that Azure gives you. You can 'environment-space' this just like you do to your `database.yml` file, so different environments can use different accounts:

    development:
      storage_account_name: foo
      storage_access_key: 123...
    test:
      storage_account_name: foo
      storage_access_key: abc...
    production:
      storage_account_name: foo
      storage_access_key: 456...

This is not required, however, and the file may simply look like this:

    storage_account_name: foo
    storage_access_key: 456...

In which case, those access keys will be used in all environments. You can also put your container name in this file, instead of adding it to the code directly. This is useful when you want the same account but a different container for development versus production.


=== Private Blob Access

In the even that are using a Blob that has been configured for Private access, you will need to use the Shared Access Signature functionality of Azure.  This functionality has been baked in to the `Attachment#expiring_url` method. Simply specify a time and a style and you will get a proper URL as follows:

    object.attachment.expiring_url(30.minutes.since, :thumb)

For more information about Azure Shared Access Signatures, please refer to [here](http://azure.microsoft.com/en-us/documentation/articles/storage-dotnet-shared-access-signature-part-1/).

=== Azure Environments

Microsoft offers specialized Azure implementations for special circumstances should the need arise.  As of the most recent update of this gem, the AzureChinaCloud, AzureUSGovernment, and AzureGermanCloud environments all offer specific storage URL's that differ from those of the standard AzureCloud.  These regions can be specified via the `:region` key of the `:azure_credentials` dictionary by using the symbols `:cn`, `:usgovt`, and `:de` respectively.  When working with one of these environments, simply update your credentials to include the region as follows:

    Paperclip::Attachment.default_options[:azure_credentials] = {
        storage_account_name: ENV['AZURE_STORAGE_ACCOUNT'],
        storage_access_key:   ENV['AZURE_STORAGE_ACCESS_KEY'],
        container:            ENV['AZURE_CONTAINER_NAME'],
        region:               :de
    }

Or, in the instance where the credentials are specified at the model level:

    has_attached_file :download,
                      storage: :azure,
                      azure_credentials: {
                        storage_account_name: ENV['AZURE_STORAGE_ACCOUNT'],
                        storage_access_key:   ENV['AZURE_STORAGE_ACCESS_KEY'],
                        container:            ENV['AZURE_CONTAINER_NAME'],
                        region:               :cn
                      }

== REQUIREMENTS:

* An Azure storage account.

== INSTALL:

Add this line to your application's Gemfile after the Paperclip gem:

    gem 'paperclip-azure', '~> 1.0'

And then execute:

    $ bundle install

== DEVELOPERS:

* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet.
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it.
* Fork the project.
* After checking out the source, run:

      $ rake newb

  This task will install any missing dependencies, run the tests/specs, and generate the RDoc.
* Start a feature/bugfix branch.
* Commit and push until you are happy with your contribution.
* Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
* Please try not to mess with the Rakefile, version, or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.
* Submit a pull request for the finished product's integration.

== LICENSE:

(The MIT License)

Copyright (c) 2017 Supportify, Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 1.0.0 / 2017-07-24

* 1 major enhancement

  * Birthday!



spec/spec_helper.rb
require 'rubygems'
require 'rspec'
require 'active_record'
require 'active_record/version'
require 'active_support'
require 'active_support/core_ext'
require 'ostruct'
require 'pathname'
require 'activerecord-import'
require 'simplecov'
require 'yaml'
require 'paperclip'

ROOT = Pathname(File.expand_path(File.join(File.dirname(__FILE__), '..')))

module SimpleCov::Configuration
  def clean_filters
    @filters = []
  end
end

SimpleCov.configure do
  clean_filters
  load_adapter 'test_frameworks'
end

ENV["COVERAGE"] && SimpleCov.start do
  add_filter "/.rvm/"
end

$LOAD_PATH << File.join(ROOT, 'lib')
$LOAD_PATH << File.join(ROOT, 'lib', 'paperclip')
require File.join(ROOT, 'lib', 'paperclip-azure.rb')

config = YAML::load(IO.read(File.dirname(__FILE__) + '/database.yml'))
ActiveRecord::Base.logger = Logger.new(File.dirname(__FILE__) + "/debug.log")
ActiveRecord::Base.establish_connection(config['test'])
if ActiveRecord::VERSION::STRING >= "4.2" &&
    ActiveRecord::VERSION::STRING < "5.0"
  ActiveRecord::Base.raise_in_transactional_callbacks = true
end
Paperclip.options[:logger] = ActiveRecord::Base.logger

Dir[File.join(ROOT, 'spec', 'support', '**', '*.rb')].each{|f| require f }

Rails = FakeRails.new('test', Pathname.new(ROOT).join('tmp'))
ActiveSupport::Deprecation.silenced = true

RSpec.configure do |config|
  config.include ModelReconstruction
  config.include TestData
  config.extend VersionHelper
  config.before(:all) do
    rebuild_model
  end
end


spec/paperclip-azure_spec.rb
require File.expand_path(File.dirname(__FILE__) + '/spec_helper')

describe "PaperclipAzure" do
  it "fails" do
    fail "hey buddy, you should probably rename this file and start specing for real"
  end
end


spec/paperclip/storage/azure_spec.rb
require 'spec_helper'
require "base64"

describe Paperclip::Storage::Azure do
  let(:storage_access_key) { 'kiaY4+GkLMVxnfOK2X+eCJOE06J8QtHC6XNuXVwt8Pp4kMezYaa7cNjtYnZr4/b732RKdz5pZwl8RN9yb8gBCg==' }

  describe "#parse_credentials" do
    let(:credentials) {{
      'production' => {key: '12345'},
      development: {key: "54321"}
    }}

    before do
      @proxy_settings = {host: "127.0.0.1", port: 8888, user: "foo", password: "bar"}
      rebuild_model storage: :azure,
                    container: "testing",
                    azure_credentials: {not: :important}
      @dummy = Dummy.new
      @avatar = @dummy.avatar
    end

    it "gets the correct credentials when RAILS_ENV is production" do
      rails_env("production") do
        expect(@avatar.parse_credentials(credentials)).to eq({key: "12345"})
      end
    end

    it "gets the correct credentials when RAILS_ENV is development" do
      rails_env("development") do
        expect(@avatar.parse_credentials(credentials)).to eq({key: "54321"})
      end
    end

    it "returns the argument if the key does not exist" do
      rails_env("not really an env") do
        expect(@avatar.parse_credentials(test: "12345")).to eq({test: "12345"})
      end
    end

  end

  describe '#container_name' do
    describe ":container option via :azure_credentials" do
      before do
        rebuild_model storage: :azure,
                      azure_credentials: {container: 'testing'}
        @dummy = Dummy.new
      end

      it "populates #container_name" do
        expect(@dummy.avatar.container_name).to eq('testing')
      end
    end

    describe ":container option" do
      before do
        rebuild_model storage: :azure,
                      container: "testing",
                      azure_credentials: {}
        @dummy = Dummy.new
      end

      it "populates #container_name" do
        expect(@dummy.avatar.container_name).to eq('testing')
      end
    end

    describe "missing :container option" do
      before do
        rebuild_model storage: :azure,
                      azure_credentials: {not: :important}

        @dummy = Dummy.new
        @dummy.avatar = stringy_file
      end

      it "raises an argument error" do
        expect{ @dummy.avatar.container_name }.to raise_error(ArgumentError, /missing required :container option/)
      end
    end
  end

  describe "" do
    before do
      rebuild_model storage: :azure,
                    azure_credentials: {
                      storage_account_name: 'storage',
                      storage_access_key: storage_access_key
                    },
                    container: "container",
                    path: ":attachment/:basename:dotextension",
                    url: ":azure_path_url"

      @dummy = Dummy.new
      @dummy.avatar = stringy_file

      allow(@dummy).to receive(:new_record?).and_return(false)
    end

    it "returns urls based on Azure paths" do
      expect(@dummy.avatar.url).to match(%r{^https://storage.blob.core.windows.net/container/avatars/data[^\.]})
    end
  end

  describe "An attachment that uses Azure for storage and has styles that return different file types" do
    before do
      rebuild_model storage: :azure,
                    styles: { large: ['500x500#', :jpg] },
                    container: "container",
                    path: ":attachment/:basename:dotextension",
                    azure_credentials: {
                      'access_key_id' => "12345",
                      'secret_access_key' => "54321"
                    }

        File.open(fixture_file('5k.png'), 'rb') do |file|
          @dummy = Dummy.new
          @dummy.avatar = file

          allow(@dummy).to receive(:new_record?).and_return(false)
        end
    end

    it "returns a url containing the correct original file mime type" do
      expect(@dummy.avatar.url).to match(/.+\/5k.png/)
    end

    it "returns a url containing the correct processed file mime type" do
      expect(@dummy.avatar.url(:large)).to match(/.+\/5k.jpg/)
    end
  end

  describe "An attachment that uses Azure for storage and has spaces in file name" do
    before do
      rebuild_model storage: :azure,
                    styles: { large: ["500x500#", :jpg] },
                    container: "container",
                    azure_credentials: {
                      "storage_access_key" => "54321"
                    }

      File.open(fixture_file("spaced file.png"), "rb") do |file|
        @dummy = Dummy.new
        @dummy.avatar = file

        allow(@dummy).to receive(:new_record?).and_return(false)
      end
    end

    it "returns a replaced version for path" do
      expect(@dummy.avatar.path).to match(/.+\/spaced_file\.png/)
    end

    it "returns a replaced version for url" do
      expect(@dummy.avatar.url).to match(/.+\/spaced_file\.png/)
    end
  end

  describe "An attachment that uses Azure for storage and has a question mark in file name" do
    before do
      rebuild_model storage: :azure,
                    styles: { large: ['500x500#', :jpg] },
                    container: "container",
                    azure_credentials: {
                      'storage_access_key' => "54321"
                    }

      stringio = stringy_file
      class << stringio
        def original_filename
          "question?mark.png"
        end
      end

      file = Paperclip.io_adapters.for(stringio)

      @dummy = Dummy.new
      @dummy.avatar = file
      @dummy.save

      allow(@dummy).to receive(:new_record?).and_return(false)
    end

    it "returns a replaced version for path" do
      expect(@dummy.avatar.path).to match(/.+\/question_mark\.png/)
    end

    it "returns a replaced version for url" do
      expect(@dummy.avatar.url).to match(/.+\/question_mark\.png/)
    end
  end

  describe ":asset_host path Interpolations" do
    before do
      rebuild_model storage: :azure,
        azure_credentials: {},
        container: "container",
        path: ":attachment/:basename:dotextension",
        url: ":asset_host"
      @dummy = Dummy.new
      @dummy.avatar = stringy_file
      allow(@dummy).to receive(:new_record?).and_return(false)
    end

    it "returns a relative URL for Rails to calculate assets host" do
      expect(@dummy.avatar.url).to match(%r{^avatars/data[^\.]})
    end
  end

  describe "#expiring_url" do
    before { @dummy = Dummy.new }

    describe "with no attachment" do
      before { expect(@dummy.avatar.exists?).to be_falsey }

      it "returns the default URL" do
        expect(@dummy.avatar.expiring_url).to eq(@dummy.avatar.url)
      end

      it 'generates a url for a style when a file does not exist' do
        expect(@dummy.avatar.expiring_url(3600, :thumb)).to eq(@dummy.avatar.url(:thumb))
      end
    end
  end

  describe "Generating a url with an expiration for each style" do
    before do
      rebuild_model storage: :azure,
                    azure_credentials: {
                      production: {
                        storage_account_name: 'prod_storage',
                        storage_access_key: 'YWNjZXNzLWtleQ==',
                        container: "prod_container"
                      },
                      development: {
                        storage_account_name: 'dev_storage',
                        storage_access_key: 'YWNjZXNzLWtleQ==',
                        container: "dev_container"
                      }
                    },
                    path: ":attachment/:style/:basename:dotextension"

      rails_env("production") do
        @dummy = Dummy.new
        @dummy.avatar = stringy_file
      end

      allow(::Azure::Storage::Core::Auth::SharedAccessSignature).to receive(:new).and_call_original
      allow(::Azure::Storage::Core::Auth::SharedAccessSignatureSigner).to receive(:new).and_call_original
    end

    it "generates a url for the thumb" do
      rails_env("production") do
        expect { @dummy.avatar.expiring_url(1800, :thumb) }.not_to raise_error
      end

      expect(::Azure::Storage::Core::Auth::SharedAccessSignature).to have_received(:new)
        .with('prod_storage', anything)
    end

    it "generates a url for the default style" do
      rails_env("production") do
        expect { @dummy.avatar.expiring_url(1800) }.not_to raise_error
      end

      expect(::Azure::Storage::Core::Auth::SharedAccessSignature).to have_received(:new)
        .with('prod_storage', anything)
    end
  end

  context "Parsing Azure credentials with a container in them" do
    before do
      rebuild_model storage: :azure,
                    azure_credentials: {
                      production: { container: "prod_container" },
                      development: { container: "dev_container" }
                    }
      @dummy = Dummy.new
    end

    it "gets the right container in production" do
      rails_env("production") do
        expect(@dummy.avatar.container_name).to eq("prod_container")
      end
    end

    it "gets the right container in development" do
      rails_env("development") do
        expect(@dummy.avatar.container_name).to eq("dev_container")
      end
    end
  end

  context "An attachment with Azure storage" do
    before do
      rebuild_model storage: :azure,
                    container: "testing",
                    path: ":attachment/:style/:basename:dotextension",
                    azure_credentials: {
                      storage_account_name: 'storage',
                      storage_access_key: storage_access_key
                    }
    end

    it "is extended by the Azure module" do
      expect(Dummy.new.avatar).to be_a(Paperclip::Storage::Azure)
    end

    it "won't be extended by the Filesystem module" do
      expect(Dummy.new.avatar).not_to be_a(Paperclip::Storage::Filesystem)
    end
    end

  describe "An attachment with Azure storage and container defined as a Proc" do
    before do
      rebuild_model storage: :azure,
                    container: lambda { |attachment| "container_#{attachment.instance.other}" },
                    azure_credentials: {not: :important}
    end

    it "gets the right container name" do
      expect(Dummy.new(other: 'a').avatar.container_name).to eq("container_a")
      expect(Dummy.new(other: 'b').avatar.container_name).to eq("container_b")
    end
  end

  context "An attachment with Azure storage and Azure credentials defined as a Proc" do
    before do
      rebuild_model storage: :azure,
                    container: {not: :important},
                    azure_credentials: lambda { |attachment|
                      Hash['storage_access_key' => "secret#{attachment.instance.other}"]
                    }
    end

    it "gets the right credentials" do
      expect(Dummy.new(other: '1234').avatar.azure_credentials[:storage_access_key]).to eq("secret1234")
    end
  end

  context "An attachment with Azure storage and Azure credentials in an unsupported manor" do
    before do
      rebuild_model storage: :azure,
                    container: "testing",
                    azure_credentials: ["unsupported"]
      @dummy = Dummy.new
    end

    it "does not accept the credentials" do
      expect { @dummy.avatar.azure_credentials }.to raise_error(ArgumentError)
    end
  end

  context "An attachment with Azure storage and Azure credentials not supplied" do
    before do
      rebuild_model storage: :azure, container: "testing"
      @dummy = Dummy.new
    end

    it "does not parse any credentials" do
      expect(@dummy.avatar.azure_credentials).to eq({})
    end
  end

  describe "with Azure credentials supplied as Pathname" do
    before do
      ENV['AZURE_CONTAINER'] = 'pathname_container'
      ENV['AZURE_STORAGE_ACCOUNT'] = 'pathname_storage_account'
      ENV['AZURE_STORAGE_ACCESS_KEY'] = storage_access_key

      rails_env('test') do
        rebuild_model storage: :azure,
          azure_credentials: Pathname.new(fixture_file('azure.yml'))

        Dummy.delete_all
        @dummy = Dummy.new
      end
    end

    it "parses the credentials" do
      expect(@dummy.avatar.container_name).to eq('pathname_container')
    end
  end

  describe "with Azure credentials in a YAML file" do
    before do
      ENV['AZURE_CONTAINER'] = 'pathname_container'
      ENV['AZURE_STORAGE_ACCOUNT'] = 'pathname_storage_account'
      ENV['AZURE_STORAGE_ACCESS_KEY'] = storage_access_key

      rails_env('test') do
        rebuild_model storage: :azure,
          azure_credentials: File.new(fixture_file('azure.yml'))

        Dummy.delete_all

        @dummy = Dummy.new
      end
    end

    it "runs the file through ERB" do
      expect(@dummy.avatar.container_name).to eq('pathname_container')
    end
  end

  describe "path is a proc" do
    before do
      rebuild_model storage: :azure,
                    path: ->(attachment) { attachment.instance.attachment_path }

      @dummy = Dummy.new
      @dummy.class_eval do
        def attachment_path
          '/some/dynamic/path'
        end
      end
      @dummy.avatar = stringy_file
    end

    it "returns a correct path" do
      expect(@dummy.avatar.path).to eq('/some/dynamic/path')
    end
  end

  private

  def rails_env(env)
    stored_env, Rails.env = Rails.env, env
    begin
      yield
    ensure
      Rails.env = stored_env
    end
  end
end


lib/azure/core/auth/shared_access_signature.rb
# Adapted from the SharedAccessSignature class found in the Azure 
# Extensions project by David Michael located at https://github.com/dmichael/azure-contrib

require 'hashie/dash'
require 'addressable/uri'

module Azure
  module Core
    module Auth
      class SharedAccessSignature
        class Version20130815 < Hashie::Dash
          property :resource,    default: 'c', required: true
          property :permissions, default: 'r', required: true
          property :start,       default: ''
          property :identifier,  default: ''
          property :expiry, required: true
          property :canonicalized_resource, required: true
          property :access_key, required: true
          # Do not change this
          property :version, default: '2103-08-15'
        end

        attr_accessor :uri, :options

        def initialize(uri, options = {}, account =  ENV['AZURE_STORAGE_ACCOUNT'])
          # This is the uri that we are signing
          @uri = Addressable::URI.parse(uri)

          is_blob = options[:resource] == 'b'

          # Create the options hash that will be turned into a query string
          @options = Version20130815.new(options.merge(canonicalized_resource: canonicalized_resource(@uri, account, is_blob)))
        end

        # Create a "canonicalized resource" from the full uri to be signed
        def canonicalized_resource(uri, account = ENV['AZURE_STORAGE_ACCOUNT'], is_blob = false)
          path = uri.path # Addressable::URI
          # There is only really one level deep for containers, the remainder is the BLOB key (that looks like a path)
          path_array = path.split('/').reject {|p| p == ''}
          container = path_array.shift

          string = if is_blob
            File.join('/', account, container, path_array.join('/'))
          else
            File.join('/', account, container)
          end

          string
        end

        # When creating the query string from the options, we only include the no-empty fields
        # - this is opposed to the string that gets signed which includes them as blank.
        def create_query_values(options = Version20130815.new)
          # Parts
          parts       = {}
          parts[:st]  = URI.unescape(options[:start]) unless options[:start] == ''
          parts[:se]  = URI.unescape(options[:expiry])
          parts[:sr]  = URI.unescape(options[:resource])
          parts[:sp]  = URI.unescape(options[:permissions])
          parts[:si]  = URI.unescape(options[:identifier]) unless options[:identifier] == ''
          parts[:sig] = URI.unescape( create_signature(options) )

          parts
        end

        def create_signature(options = Version20130815)
          string_to_sign  = []
          string_to_sign << options[:permissions]
          string_to_sign << options[:start]
          string_to_sign << options[:expiry]
          string_to_sign << options[:canonicalized_resource]
          string_to_sign << options[:identifier]

          ::Azure::Core::Auth::Signer.new(options[:access_key]).sign(string_to_sign.join("\n").force_encoding("UTF-8"))
        end

        def sign
          @uri.query_values = create_query_values(@options)
          @uri.to_s
        end
      end
    end
  end
end

lib/paperclip/storage/azure.rb
require 'azure/storage'
require 'paperclip/storage/azure/environment'

module Paperclip
  module Storage
    # Azure's container file hosting service is a scalable, easy place to store files for
    # distribution. You can find out more about it at http://azure.microsoft.com/en-us/services/storage/
    #
    # There are a few Azure-specific options for has_attached_file:
    # * +azure_credentials+: Takes a path, a File, a Hash or a Proc. The path (or File) must point
    #   to a YAML file containing the +storage_access_key+ and +storage_account_name+ that azure
    #   gives you. You can 'environment-space' this just like you do to your
    #   database.yml file, so different environments can use different accounts:
    #     development:
    #       storage_account_name: foo
    #       storage_access_key: 123...
    #     test:
    #       storage_account_name: foo
    #       storage_access_key: abc...
    #     production:
    #       storage_account_name: foo
    #       storage_access_key: 456...
    #   This is not required, however, and the file may simply look like this:
    #     storage_account_name: foo
    #     storage_access_key: 456...
    #   In which case, those access keys will be used in all environments. You can also
    #   put your container name in this file, instead of adding it to the code directly.
    #   This is useful when you want the same account but a different container for
    #   development versus production.
    #   When using a Proc it provides a single parameter which is the attachment itself. A
    #   method #instance is available on the attachment which will take you back to your
    #   code. eg.
    #     class User
    #       has_attached_file :download,
    #                         :storage => :azure,
    #                         :azure_credentials => Proc.new{|a| a.instance.azure_credentials }
    #
    #       def azure_credentials
    #         { :container => "xxx", :storage_account_name => "xxx", :storage_access_key => "xxx" }
    #       end
    #     end
    #
    # * +container+: This is the name of the Azure container that will store your files. Remember
    #   that the container must be unique across the storage account. If the container does not exist
    #   Paperclip will attempt to create it. The container name will not be interpolated.
    #   You can define the container as a Proc if you want to determine it's name at runtime.
    #   Paperclip will call that Proc with attachment as the only argument.
    # * +path+: This is the key under the container in which the file will be stored. The
    #   URL will be constructed from the container and the path. This is what you will want
    #   to interpolate. Keys should be unique, like filenames, and despite the fact that
    #   Azure (strictly speaking) does not support directories, you can still use a / to
    #   separate parts of your file name.
    # * +region+: Depending on the region, different base urls are used. Supported values :global, :de

    module Azure
      def self.extended base
        begin
          require 'azure'
        rescue LoadError => e
          e.message << " (You may need to install the azure SDK gem)"
          raise e
        end unless defined?(::Azure::Core)

        base.instance_eval do
          @azure_options     = @options[:azure_options]     || {}

          unless @options[:url].to_s.match(/\A:azure.*url\z/) || @options[:url] == ":asset_host".freeze
            @options[:path] = path_option.gsub(/:url/, @options[:url]).sub(/\A:rails_root\/public\/system/, "".freeze)
            @options[:url]  = ":azure_path_url".freeze
          end
          @options[:url] = @options[:url].inspect if @options[:url].is_a?(Symbol)

          @http_proxy = @options[:http_proxy] || nil
        end

        Paperclip.interpolates(:azure_path_url) do |attachment, style|
          attachment.azure_uri(style)
        end unless Paperclip::Interpolations.respond_to? :azure_path_url
        Paperclip.interpolates(:asset_host) do |attachment, style|
          "#{attachment.path(style).sub(%r{\A/}, "".freeze)}"
        end unless Paperclip::Interpolations.respond_to? :asset_host
      end

      def expiring_url(time = 3600, style_name = default_style)
        if path(style_name)
          uri = URI azure_uri(style_name)
          generator = ::Azure::Storage::Core::Auth::SharedAccessSignature.new azure_account_name,
                                                                              azure_credentials[:storage_access_key]

          generator.signed_uri uri, false, service:      'b',
                                           resource:     'b',
                                           permissions:  'r',
                                           start:        (Time.now - (5 * 60)).utc.iso8601,
                                           expiry:       (Time.now + time).utc.iso8601
        else
          url(style_name)
        end
      end

      def auto_connect_duration
        @auto_connect_duration ||= @options[:auto_connect_duration] || azure_credentials[:auto_connect_duration] || 10
        @auto_connect_duration
      end

      def azure_credentials
        @azure_credentials ||= parse_credentials(@options[:azure_credentials])
      end

      def azure_account_name
        account_name = @options[:azure_storage_account_name] || azure_credentials[:storage_account_name]
        account_name = account_name.call(self) if account_name.is_a?(Proc)

        account_name
      end

      def container_name
        @container ||= @options[:container] || azure_credentials[:container]
        @container = @container.call(self) if @container.respond_to?(:call)
        @container or raise ArgumentError, "missing required :container option"
      end

      def azure_interface
        @azure_interface ||= begin
          config = {}

          [:storage_account_name, :storage_access_key, :container].each do |opt|
            config[opt] = azure_credentials[opt] if azure_credentials[opt]
          end

          obtain_azure_instance_for(config.merge(@azure_options))
        end
      end

      def azure_storage_client
        config = {}

        [:storage_account_name, :storage_access_key].each do |opt|
          config[opt] = azure_credentials[opt] if azure_credentials[opt]
        end

        @azure_storage_client ||= ::Azure::Storage::Client.create config
      end

      def obtain_azure_instance_for(options)
        instances = (Thread.current[:paperclip_azure_instances] ||= {})
        return instances[options] if instance[options]

        service = ::Azure::Storage::Blob::BlobService.new(client: azure_storage_client)
        service.with_filter ::Azure::Storage::Core::Filter::ExponentialRetryPolicyFilter.new

        instances[options] = service
      end

      def azure_uri(style_name = default_style)
        "https://#{azure_base_url}/#{container_name}/#{path(style_name).gsub(%r{\A/}, '')}"
      end

      def azure_base_url
        Environment.url_for azure_account_name, azure_credentials[:region]
      end

      def azure_container
        @azure_container ||= azure_interface.get_container_properties container_name
      end

      def azure_object(style_name = default_style)
        azure_interface.get_blob_properties container_name, path(style_name).sub(%r{\A/},'')
      end

      def parse_credentials(creds)
        creds = creds.respond_to?('call') ? creds.call(self) : creds
        creds = find_credentials(creds).stringify_keys
        env = Object.const_defined?(:Rails) ? Rails.env : nil
        (creds[env] || creds).symbolize_keys
      end

      def exists?(style = default_style)
        if original_filename
          !azure_object(style).nil?
        else
          false
        end
      rescue ::Azure::Core::Http::HTTPError => e
        raise unless e.status_code == 404

        false
      end

      def create_container
        azure_interface.create_container container_name
      end

      def flush_writes #:nodoc:
        @queued_for_write.each do |style, file|
          retries = 0
          begin
            log("saving #{path(style)}")

            write_options = {
              content_type: file.content_type,
            }

            if azure_container
              save_blob container_name, path(style).sub(%r{\A/},''), file, write_options
            end
          rescue ::Azure::Core::Http::HTTPError => e
            if e.status_code == 404
              create_container
              retry
            else
              raise
            end
          ensure
            file.rewind
          end
        end

        after_flush_writes # allows attachment to clean up temp files

        @queued_for_write = {}
      end

      def save_blob(container_name, storage_path, file, write_options)

        if file.size < 64.megabytes
          azure_interface.create_block_blob container_name, storage_path, file.read, write_options
        else
          blocks = []; count = 0
          while data = file.read(4.megabytes)
            block_id = "block_#{(count += 1).to_s.rjust(5, '0')}"

            azure_interface.create_blob_block container_name, storage_path, block_id, data

            blocks << [block_id]
          end

          azure_interface.commit_blob_blocks container_name, storage_path, blocks
        end
      end

      def flush_deletes #:nodoc:
        @queued_for_delete.each do |path|
          begin
            log("deleting #{path}")

            azure_interface.delete_blob container_name, path
          rescue ::Azure::Core::Http::HTTPError => e
            raise unless e.status_code == 404
          end
        end
        @queued_for_delete = []
      end

      def copy_to_local_file(style, local_dest_path)
        log("copying #{path(style)} to local file #{local_dest_path}")

        blob, content = azure_interface.get_blob(container_name, path(style).sub(%r{\A/},''))

        ::File.open(local_dest_path, 'wb') do |local_file|
          local_file.write(content)
        end
      rescue ::Azure::Core::Http::HTTPError => e
        raise unless e.status_code == 404

        warn("#{e} - cannot copy #{path(style)} to local file #{local_dest_path}")
        false
      end

      private

      def find_credentials creds
        case creds
        when File
          YAML::load(ERB.new(File.read(creds.path)).result)
        when String, Pathname
          YAML::load(ERB.new(File.read(creds)).result)
        when Hash
          creds
        when NilClass
          {}
        else
          raise ArgumentError, "Credentials given are not a path, file, proc, or hash."
        end
      end
    end
  end
end
